### PR TITLE
Error logging for 5xx and -1 downstream responses, unit test improvements

### PR DESCRIPTION
**Pull request by wjhoward**
The error log message will look similar to the following (note I temporarily adjusted the code to log a 4xx message for the output below):

```
2021/09/14 05:42:28 [error] 22449#22449: *2 [lua] spectre_common.lua:36: log():

 {"critical":false,"err":"X-Original-Status: 404, message: <error message here>"},

client: 127.0.0.1, server: , request: "GET /category_yel HTTP/1.1", host: "internalapi"
```
I did intend to update the unit tests to check that the log function was being called, but I ran into some suspected scoping issues when using the busted `spy` methods. That being said, test coverage is running over the new lines of code:

```
              -- Log any "no_response" errors
    1         local log_error_message = HEADERS.ORIGINAL_STATUS .. ": -1, message: " .. error_message
    1         log(ngx.ERR, { err=log_error_message, critical=false })
    ....
              -- Log any 5xx errors
    2     if string.match(response.status, '5%d%d') then
    1         local log_error_message = HEADERS.ORIGINAL_STATUS .. ": " .. response.status .. ", message: " .. error_message
    1         log(ngx.ERR, { err=log_error_message, critical=false })
```

**Unit tests (coverage slightly improved):**
```
Total: 0 warnings / 0 errors in 14 files
./luawrapper luacov
==============================================================================

File                      Hits Missed Coverage
----------------------------------------------
lua/bulk_endpoints.lua    10   95     9.52%
lua/caching_handlers.lua  149  2      98.68%
lua/config_loader.lua     73   7      91.25%
lua/datastores.lua        202  11     94.84%
lua/http.lua              5    8      38.46%
lua/internal_handlers.lua 30   64     31.91%
lua/metrics_helper.lua    50   34     59.52%
lua/spectre_common.lua    263  41     86.51%
lua/traceback.lua         6    0      100.00%
lua/util.lua              26   0      100.00%
lua/zipkin.lua            28   38     42.42%
----------------------------------------------
Total                     842  300    73.73%
```

### CODE REVIEW

File: lua/spectre_common.lua [hunk 1]
```diff
@@ -312,6 +312,11 @@ local function forward_to_destination(method, request_uri, request_headers)
             message = error_message,
             status_code = status,
         })
+
+        -- Log any "no_response" errors
+        local log_error_message = HEADERS.ORIGINAL_STATUS .. ": -1, message: " .. error_message
+        log(ngx.ERR, { err=log_error_message, critical=false })
```
Comment (sarguru) [inline lua/spectre_common.lua:318]: This is probably worth tagging as critical?

File: tests/lua/spectre_common_test.lua [hunk 2]
```diff
@@ -646,18 +650,19 @@ describe("spectre_common", function()
...
 
-            assert.are.equal(500, resp.status)
-            assert.are.equal('Error requesting /bar: there was an error. Whoops.', resp.body)
+            assert.are.equal(200, resp.status)
+            assert.are.equal(false, resp.no_response)
+            assert.are.equal('RESULT', resp.body)
         end)
 
-        it("Forwards body when HTTP request succeeds", function()
+        it("Errors out when the downstream HTTP request returns an error", function()
```
Comment (sarguru) [inline tests/lua/spectre_common_test.lua:658]: Do you think its worth covering the -1 case as well here?

**Review by sarguru — CHANGES_REQUESTED** [review]

File: tests/lua/spectre_common_test.lua (hunk 2 above)
Comment (wjhoward) [reply tests/lua/spectre_common_test.lua:658]: So the following test covers the "no response" scenario, which triggers the -1 log / header:
https://github.com/Yelp/casper/blob/98ad7e145831382120bb6f3f7c1342ec6fa493c9/tests/lua/spectre_common_test.lua#L681

Comment (sarguru) [reply tests/lua/spectre_common_test.lua:658]: 👍

**Pull request merged**